File: Android/TCDemo/app/src/IABDebug/java/com/tagcommander/tcdemo/TCExample/TCServerSideImplementation.kt
package com.tagcommander.tcdemo.TCExample

import android.content.Context
import android.provider.Settings
import android.util.Log
import com.tagcommander.lib.core.TCDebug
import com.tagcommander.lib.serverside.TCServerSide
import com.tagcommander.lib.serverside.events.TCAddPaymentInfoEvent
import com.tagcommander.lib.serverside.events.TCAddShippingInfoEvent
import com.tagcommander.lib.serverside.events.TCAddToCartEvent
import com.tagcommander.lib.serverside.events.TCBeginCheckoutEvent
import com.tagcommander.lib.serverside.events.TCCustomEvent
import com.tagcommander.lib.serverside.events.TCRemoveFromCartEvent
import com.tagcommander.lib.serverside.schemas.TCItem
import com.tagcommander.lib.serverside.schemas.TCProduct
import com.tagcommander.tcdemo.models.TCModel
import com.tagcommander.tcdemo.ui.theme.Product
import com.tagcommander.tcdemo.BuildConfig

object TCServerSideImplementation {

    val siteID = 4183
    val sourceKey = "939324f7-28ec-4352-9241-ed4ca8405b07"
    lateinit var tc: TCServerSide
    private lateinit var appContext: Context

    fun initTCServerSide(context: Context)
    {
        tc = TCServerSide(siteID, sourceKey, context)
        appContext = context
    }

    fun sendPageViewEvent(pageName:String, pageType:String)
    {
        val event = TCCustomEvent("start")
        event.addAdditionalProperty("custom_page_key", pageName)
        event.addAdditionalProperty("custom_page_type", pageType)
        event.addAdditionalProperty("auto_timezone", Settings.Global.getInt(appContext.getContentResolver(), Settings.Global.AUTO_TIME_ZONE))
        event.addAdditionalProperty("auto_time", Settings.Global.getInt(appContext.getContentResolver(), Settings.Global.AUTO_TIME))
        tc.execute(event)
    }

    fun sendBeginCheckoutEvent()
    {
        val event = TCBeginCheckoutEvent()
        tc.execute(event)
    }

    fun sendAddToCartEvent(product: Product?)
    {
        val item = TCItem(product?.id, TCProduct(product?.id, product?.name, product?.price), 1)
        val event = TCAddToCartEvent(1.0f,  "euro", listOf(item))
        tc.execute(event)
    }

    fun sendRemoveFromCartEvent(product: Product?)
    {
        val item = TCItem(product?.id, TCProduct(product?.id, product?.name, product?.price), 1)
        val event = TCRemoveFromCartEvent(listOf(item))
        tc.execute(event)
    }

    fun sendAddPaiementInfo(method: String, coupon: String)
    {
        val cartItems = TCModel.products.filter { it.isInCart }.map { TCItem(it.id, TCProduct(it.id, it.name, it.price), 1) }
        val event = TCAddPaymentInfoEvent("card")
        event.coupon = coupon
        event.currency = "EUR"
        event.items = cartItems
        event.addAdditionalProperty("method", method)
        tc.execute(event)
    }

    fun sendAddShippingInfo(adress:String)
    {
        val event = TCAddShippingInfoEvent()
        event.addAdditionalProperty("adress", adress)
        tc.execute(event)
    }

    fun sendPurchaseEvent()
    {
        val event = TCAddShippingInfoEvent()
        tc.execute(event)
    }
}

File: iOS/TCDemo/TCExample/TCServerSideImplementation.swift
//
//  TCServerSideImplementation.swift
//  TCDemo
//
//  Created by Abdelhakim SAID on 13/10/2025.
//
import TCCore
import TCServerSide_noIDFA
import UserNotifications

struct TCServerSideImplementation {
    private static let siteID: Int32 = 3311
    private static let sourceKey = "632a2280-880f-4812-a311-4c5e07cf90e9"

    private static var tc: ServerSide? = nil

    static func initTCServerSide() {
        #if TC_USE_SERVER_SIDE
        if let firebaseInstance = TCFirebaseImplementation.getInstance() { // only if you use Firebase event forwarding
            tc = ServerSide(
                siteID: siteID,
                andSourceKey: sourceKey,
                andDefaultBehaviour: PB_ALWAYS_ENABLED,
                andFirebaseInstance: firebaseInstance
            )
        } else {
            tc = ServerSide(            // classic initialisation otherwise
                siteID: siteID,
                andSourceKey: sourceKey,
                andDefaultBehaviour: PB_ALWAYS_ENABLED
            )
        }
        #else
        Toast.show(message: "Consent only ! ServerSide is not available on this App Target")
        #endif
    }


    static func sendPageViewEvent(pageName: String, pageType: String) {
        let event = TCPageViewEvent(type: pageType)
        event?.pageName = pageName
        event?.addAdditionalProperty("test_code", withStringValue: "test_code")
        tc?.execute(event)
    }

    static func sendBeginCheckoutEvent() {
        let event = TCBeginCheckoutEvent()
        tc?.execute(event)
    }

    static func sendAddToCartEvent(product: Product) {
        let item: TCItem = TCItem(
            itemId: product.id,
            with: TCProduct(
                productId: product.id,
                withName: product.name,
                withPrice: NSDecimalNumber(value: product.price)
            ),
            withQuantity: 1
        )
        let event = TCAddToCartEvent(
            value: 1,
            withCurrency: "euro",
            withItems: [item]
        )
        tc?.execute(event)
    }

    static func sendAddToWishListEvent(product: Product) {
        let item = TCItem(
            itemId: product.id,
            with: TCProduct(
                productId: product.id,
                withName: product.name,
                withPrice: NSDecimalNumber(value: product.price)
            ),
            withQuantity: 1
        )
        let event = TCAddToWishlistEvent(items: [item!])
        tc?.execute(event)
    }

    static func sendRemoveFromCartEvent(product: Product) {
        let item = TCItem(
            itemId: product.id,
            with: TCProduct(
                productId: product.id,
                withName: product.name,
                withPrice: NSDecimalNumber(value: product.price)
            ),
            withQuantity: 1
        )
        let event = TCRemoveFromCartEvent(items: [item as Any])
        tc?.execute(event)
    }

    static func sendAddPaiementInfo(
        method: String,
        coupon: String,
        product: Product
    ) {
        let item = TCItem(
            itemId: product.id,
            with: TCProduct(
                productId: product.id,
                withName: product.name,
                withPrice: NSDecimalNumber(value: product.price)
            ),
            withQuantity: 1
        )
        let event = TCAddPaymentInfoEvent(paymentMethod: "card")
        event?.coupon = coupon
        event?.currency = "EUR"
        event?.items = [item!]
        tc?.execute(event)
    }

    static func sendAddShippingInfo(adress: String) {
        let event = TCAddShippingInfoEvent()
        tc?.execute(event)
    }

    static func sendPurchaseEvent() {
        let event = TCAddShippingInfoEvent()
        tc?.execute(event)
    }

    static func getServerSideInstance() -> ServerSide? {
        return tc
    }
}
